p33113/s263068/lab1/main.c

```diff
@@ -0,0 +1,345 @@
+#include <stdio.h>
+#include <sys/mman.h>
+#include <zconf.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <semaphore.h>
+#include <fcntl.h>
+
+#define A 68
+#define B 0xDD95BF70
+#define C mmap
+#define D 120
+#define E 51
+#define F block
+#define G 96
+#define H random
+#define I 88
+#define J max
+#define K sem
+
+typedef struct {
+    unsigned char rand_fd;
+    int start_ptr;
+    int end_ptr;
+    unsigned char *mem_ptr;
+} mem_write_meta;
+
+typedef struct {
+    size_t f_amount;
+    size_t f_chunk_size;
+    size_t f_chunk_tail;
+    unsigned char *mem_ptr;
+    int *f_ds;
+} file_write_meta;
+
+typedef struct {
+    int f_d;
+    int index;
+} file_read_meta;
+
+pthread_t t_file_write_meta;
+
+sem_t semaphore;
+unsigned char *mem_ptr;
+int bytes;
+int rand_fd;
+
+int r_index;
+char r_data[10000];
+
+int f_amount;
+int *f_ds;
+
+void pause_() {
+    printf("Press enter to continue...\n");
+    fflush(stdout);
+    int c;
+    while ((c = getchar()) != 0x0a);
+    if (c != EOF) {
+        return;
+    }
+    if (feof(stdin)) {
+        exit(0);
+    }
+}
+
+void mem_preallocate() {
+    bytes = A * 1024 * 1024;
+    rand_fd = open("/dev/urandom", O_RDONLY);
+    mem_ptr = mmap(
+            (void *) B,
+            bytes,
+            PROT_READ | PROT_WRITE,
+            MAP_SHARED | MAP_ANONYMOUS,
+            -1,
+            0);
+
+    if (mem_ptr == MAP_FAILED) {
+        perror("Can't map a file\n");
+        exit(-1);
+    }
+}
+
+unsigned char r_read(int r_data_fd) {
+    if (r_index < 10000) {
+        r_index++;
+        return r_data[r_index];
+    } else {
+        int result = read(r_data_fd, &r_data, sizeof(r_data));
+        if (result == -1) {
+            perror("Can't read int from /dev/urandom\n");
+            return 0;
+        }
+        r_index = 0;
+        return r_data[r_index];
+    }
+}
+
+void *t_mem_write(void *t_mw_meta) {
+    mem_write_meta *mw_meta = (mem_write_meta *) t_mw_meta;
+    unsigned int bytes_done = 0;
+    for (int i = mw_meta->start_ptr; i < mw_meta->end_ptr; i++) {
+        unsigned char r = r_read(mw_meta->rand_fd);
+        mw_meta->mem_ptr[i] = r;
+        bytes_done++;
+    }
+    printf("Thread-#%d finished writing %uB\n", (int) pthread_self(), bytes_done);
+    free(mw_meta);
+    return NULL;
+}
+
+void mem_write() {
+    const int chunk = bytes / D / sizeof(*mem_ptr);
+    const int tail = bytes % D / sizeof(*mem_ptr);
+
+    printf("Chunk size: %dB\n", chunk);
+    printf("Chunk tail size: %dB\n", tail);
+
+    pthread_t pthread_pool[D];
+    mem_write_meta *mw_meta;
+    int thread_counter = 0;
+    for (int i = 0; i < D;) {
+        mw_meta = malloc(sizeof(*mw_meta));
```

Comment:
Давайте сделаем выделение и освобождение одним чанком?

Reply:
```suggestion
        mem_write_meta *mw_meta = malloc(D * sizeof(*mw_meta));
```
Внутри цикла теперь:
```
   for (int i = 0; i < D;) {
        mw_meta[thread_counter].mem_ptr = mem_ptr;
        mw_meta[thread_counter].rand_fd = rand_fd;
        mw_meta[thread_counter].start_ptr = i * chunk;
        mw_meta[thread_counter].end_ptr = ++i * chunk;
        if (pthread_create(&pthread_pool[thread_counter], 0, t_mem_write, (void *) mw_meta)) {
            printf("Can't create Thread\n");
        }
        thread_counter++;
    }
```
И вот здесь немного меняем:
```
if (tail != 0) {
        mw_meta->mem_ptr = mem_ptr;
        mw_meta->rand_fd = rand_fd;
        mw_meta->start_ptr = D * chunk;
        mw_meta->end_ptr = D * chunk + tail;
        t_mem_write((void *) mw_meta);
    }
```